### DISCUSSION TITLE
Register custom turbo-stream plugin for encoding/decoding additional types in loader data

### DISCUSSION THREAD

**Discussion by robmonie** [remix-run/remix#10128] [Proposals]
Edit: I've converted this to a proposal from Q&A after doing some more digging and watching the latest roadmap video covering type-safe routing in React Router V7.

I'd like to propose an api for registering custom `turbo-stream `plugins for use in the encoding / decoding of loader-data.  A specific use-case is to be able to send and receive  Luxon `DateTime` or `BigNumber` instances which we've standardised on within our apps. Currently we do this via wrappers around the existing `useDataLoader` and friends but the move to turbo-stream makes this feel a bit less natural and continuing as we currently do would mean missing some of the router improvements.

An example of registering a plugin for encoding/decoding a Luxon `DateTime`. 

```typescript
registerTurboStreamPlugin({ 
  encoder: (value) => {
    if (value instanceof DateTime) {
      return ['DateTime"' value.toISODate()];
    }
  },
  decoder: (type, encodedValue) => {
    if (type === 'DateTime') {
      return { value: DateTime.fromIso(encodedValue); };
    }
  }
});

```

**Comment by mc-dsk** [comment]
It would be amazing !

**Comment by robmonie** [comment]
Note - I've moved/copied this into the React Router Discussion in the hope it will get more traction there now that V7 is out - https://github.com/remix-run/react-router/discussions/12874
